File: ruby/problem_389.rb
require "pry"
# require "mathn"

Array.class_eval do
  # these are *TOTALLY* not-safe!!!
  def sum
    @sum ||= self.reduce(:+)
  end
  def mean
    @mean ||= sum*1.0 / self.length
  end
  def variance
    @variance ||= (self.inject(0) {|sum, n| sum + (n - self.mean)**2 }) / self.length
  end
  def merge ary
    # hopefully ary < self
    self.map { |s|
      ary.map { |a| a + s }
    }.flatten
  end
end



@@fact=[0,1]
def fact n
  @@fact[n+1] ||= n * fact(n-1)
end

def combination n, k
  (fact(n)) / (fact(k) * fact(n-k) * 1.0)
end


# An unbiased single 4-sided die is thrown and its value, T, is noted.
# T unbiased 6-sided dice are thrown and their scores are added together. The sum, C, is noted.
# C unbiased 8-sided dice are thrown and their scores are added together. The sum, O, is noted.
# O unbiased 12-sided dice are thrown and their scores are added together. The sum, D, is noted.
# D unbiased 20-sided dice are thrown and their scores are added together. The sum, I, is noted.
# Find the variance of I, and give your answer rounded to 4 decimal places.


class Dice
  def initialize num, sides
    @sides, @num = sides, num
  end

  def mean
    (1 + @sides) * @num / 2.0
  end

  def variance
    ((@sides**2)/12.0 - 1/12.0) * @num
  end

  def std_dev
    variance**(1/2.0)
  end

  def to_a
    ary=[]
    (@sides * @num - @num + 1).times {|i| ary << i + @num }
    ary
  end

  def num_prob num
    sum=0
    0.upto(((num - @num)/@sides).floor) do |i|
      sum += (-1.0)**i * combination(@num, i) * combination( (num - @sides*i - 1), (@num - 1) )
    end
    (1.0 / @sides**@num) * sum
  end

  def num
    @sides ** @num
  end

  def full_a
    ary=[]
    0.upto(@num) {|i|
      ary[i] = []
      @sides.times {|k| ary[i][k] = i+1 * k+1}
    }
    ary.flatten
  end

end

class MultiDice
  def initialize *dice
    
  end

  def mean
    
  end

end

class MultiStat
  def initialize *args
    @objs = args
  end
  
  def mean
    @mean ||= @objs.inject(0) { |mem, var|
      mem + var.mean*var.num
    } / @objs.inject(0) { |mem, var| mem + var.num }
  end

  def variance
    @variance ||= (@objs.inject(0) { |mem, var|
          mem + var.num*(var.variance + var.mean**2)
        } / @objs.inject(0) { |mem, var| mem + var.num }) - mean**2
  end
  
end

f = OpenStruct.new( mean: 65, variance: 4, num: 12)
m = OpenStruct.new( mean: 70, variance: 9, num: 12)
b = MultiStat.new m, f


@one_d2 = Dice.new 1, 2

@one_d3 = Dice.new 1, 3
@two_d3 = Dice.new 2, 3

# @two_d6 = Dice.new 2, 6

# brute forcing this thing....
@d4  = [1,2,3,4]                      # t
@d6  = [1,2,3,4,5,6]                  # c
@d8  = [1,2,3,4,5,6,7,8]              # o
@d12 = [1,2,3,4,5,6,7,8,9,10,11,12]   # d
@d20 = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]

@big_ary=[]
@d4.each do |nd4|
  @big_ary << nd4
  nd4.times { |t|

  }
end

binding.pry

File: haskell/problem_389.hs
--An unbiased single 4-sided die is thrown and its value, T, is noted.
--T unbiased 6-sided dice are thrown and their scores are added together. The sum, C, is noted.
--C unbiased 8-sided dice are thrown and their scores are added together. The sum, O, is noted.
--O unbiased 12-sided dice are thrown and their scores are added together. The sum, D, is noted.
--D unbiased 20-sided dice are thrown and their scores are added together. The sum, I, is noted.
--Find the variance of I, and give your answer rounded to 4 decimal places.

import Data.Ratio
import Data.List

type Probability = Rational
type Density a = [(a, Probability)]

triple :: [a] -> [(a, Probability)]
triple as = map (\a -> (a, 3/n)) as where  n = fromIntegral (length as)

deterministic :: a -> Density a
deterministic a = [(a, 1)]

uniform :: [a] -> [(a, Probability)]
uniform as = map (\a -> (a, 1/n)) as where n = fromIntegral (length as)
d n = uniform [1..n]

foo :: (a -> Density b) -> (Density a -> Density b)
foo f da = concat (map (\(a, pa) -> map (\(b, pb) -> (b, pa*pb)) (f a)) da)

-- foo (\x -> deterministic (x*3)) (d 6)
fmap' :: (a -> b) -> (Density a -> Density b)
fmap' fa fb = foo (\x -> deterministic (fa x)) fb

--01:11 < dmwit> Or, as an alternate slightly harder problem, try to explain what "foo d (d 6)" does. =)
--01:25 < serialhex> ok, smaller #'s make life easier: `foo d (d 3)` first samples the d3, makes another die 
--                   with the roll on the die (a d1) and multiplies it by that...  then it does the same with 
--                   the next roll: a 2, and multiplies that out, giving (2,1 % 3) * [(1,1 % 2),(2,1 % 2)] = 
--                   [(1,1 % 6),(2,1 % 6)]; and so on...
--03:12 < serialhex> foo (\x -> fmap' (\y -> x+y ) (d 2) ) (d 2)

twoDn :: (Enum b, Num b) => b -> Density b
twoDn n = foo (\x -> fmap' (\y -> x+y ) (d n) ) (d n)

norm :: Ord a => Density a -> Density a
norm a  = merge lst
        where lst = sort a
              merge (x:y:rst) = if (fst x) == (fst y)
                                    then merge (((fst x), (snd x) + (snd y)):rst)
                                    else x:merge (y:rst)
              merge anythingelse  = anythingelse

--replicateM :: Int -> Density a -> Density [a]
--replicateM n m = sequence (replicate n m)


--04:20 < serialhex> so `sequence [ d 2, d 2]` should give...
--04:22 < dmwit> [([1,1],1 % 4),([1,2],1 % 4),([2,1],1 % 4),([2,2],1 % 4)]

--foo (\x -> fmap' (\y -> [y,x] ) (d 2) ) (d 2)  works!!!  now: extend!
--sequence' :: [Density a] -> Density a
--sequence' []     = []
--sequence' (a:as) = foo (\x -> fmap' (\y -> [y,x] ) a ) (sequence as)

--23:30 < dmwit> Any function that you would write as "foo [] = z; foo (x:xs) = bar x (foo xs)" can just be "foo 
--               = foldr bar z" instead.
--23:30 < dmwit> That's all there is to it.


--pair :: Density d -> Density d -> Density d
--pair a b = concat (map (\(x, px) -> map (\(y, py) -> (x:y, px*py) ) b) a )
